zkvm/src/ops.rs

```diff
@@ -399,4 +407,61 @@ impl Program {
         self.0.push(Instruction::Cloak(m, n));
         self
     }
+
+    /// Takes predicate and closure to add choose operations for
+    /// predicate tree traversal.
+    pub fn choose_predicate<F>(
+        &mut self,
+        pred: Predicate,
+        choose_fn: F,
+    ) -> Result<&mut Program, VMError>
+    where
+        F: FnOnce(PredicateTree) -> Result<&mut PredicateTree, VMError>,
+    {
+        choose_fn(PredicateTree {
+            prog: self,
+            pred: pred,
+        })?;
+        Ok(self)
+    }
+}
+
+/// Adds data and instructions to traverse a predicate tree.
+pub struct PredicateTree<'a> {
+    prog: &'a mut Program,
+    pred: Predicate,
+}
+
+impl<'a> PredicateTree<'a> {
+    /// Left Predicate branch
+    pub fn left(&mut self) -> Result<&mut Self, VMError> {
+        let (l, r) = self.pred.to_disjunction()?;
```

Comment:
i guess this answers my question about move/borrow. That's tricky since we cannot move out of the field, even if we immediately put something back in. And the thing we want to replace this with is not sitting somewhere separately (could use mem::replace then), but it's referenced by the field self.pred itself.

Reply:
We can overcome this by making PredicateTree consumed instead of &mut-borrowed in this API. This will also help with `call()`.

```rust
pub fn left(self) -> Result<Self, VMError> {
    let (l, r) = self.pred.to_disjunction()?; // assuming self.pred is moved out
    let prog = self.prog;
    prog.push(l).push(r).left();
    Ok(Self{pred: l, prog})  // move left predicate into new instance of the PredicateTree
}
```

Reply:
Refactored to have the `PredicateTree` consumed, and expect the closure to return some `&mut Program` (rather than `&mut PredicateTree`)!